### ISSUE TITLE
pageserver: panic in memorize_materialized_page (get_vectored_impl)

### ISSUE THREAD

**Issue by petuhovskiy** [neondatabase/neon#7874]
Sentry caught a panic in pageserver, caused by `assert!(inner.buf == img);`

## Stacktrace

```
tracing::instrument::Instrumented<T>::poll (instrument.rs:272)
pageserver::basebackup::Basebackup<T>::send_tarball::{{closure}} (basebackup.rs:302)
pageserver::tenant::timeline::Timeline::get_vectored::{{closure}} (timeline.rs:1032)
pageserver::tenant::timeline::Timeline::get_vectored_impl::{{closure}} (timeline.rs:1225)
pageserver::tenant::timeline::Timeline::reconstruct_value::{{closure}} (timeline.rs:4926)
pageserver::page_cache::PageCache::memorize_materialized_page::{{closure}} (page_cache.rs:481) <= assert
core::panicking::panic (panicking.rs:145)
core::panicking::panic_fmt (panicking.rs:72)
rust_begin_unwind (panicking.rs:645)
```

Failed assert that checks content in page_cache:

https://github.com/neondatabase/neon/blob/1455f5a2612a95d1c8fe3f68311cb42fc4785523/pageserver/src/page_cache.rs#L477-L482

## Logs, links
- sentry page: https://neondatabase.sentry.io/issues/5396548687/?alert_rule_id=13489026&alert_type=issue&notification_uuid=85ebb7d9-ae3c-49f2-a57f-98804b9d1c6e&project=4504220031582208&referrer=slack
- slack notification: https://neondb.slack.com/archives/C0603MG2B0W/p1716485418100239

**Comment by skyzh** [comment]
I said the below thing might be related during the meeting and now I realized that it's probably not related 🤣

https://github.com/neondatabase/neon/blob/b52e31c1a42e186d578975cce632bf244c5f2957/pageserver/src/tenant/timeline.rs#L916-L921 as part of https://github.com/neondatabase/neon/pull/7517

**Comment by jcsp** [comment]
Similar general area of code as https://github.com/neondatabase/neon/issues/6967

**Comment by jcsp** [comment]
Let's dig into this next week & figure out if recent changes might have removed the code path that touched the page cache here

**Comment by VladLazar** [comment]
I think it's too late to do a proper investigation of this. We do not have logs since it happened more than 30 days ago.
Hence, we don't know the tenant, timeline and lsn of the basebackup request that caused this.

This could have been caused by the same underlying issue in https://github.com/neondatabase/neon/issues/6967 (fixed in https://github.com/neondatabase/neon/pull/8050). Vectored get for single page requests was rolled out to eu-central the week of 20 May and the panic happened on 23 May. Here's how it would "work":
1. The impacted tenant has a layer map layout like the one described in https://github.com/neondatabase/neon/pull/8050
2. Before the basebackup, compute issues requests for a single slru segment twice
3. For the first request, there's nothing in the cache, so we handle it correctly
4. For the second request, we have a cached image and run into the bug fixed in https://github.com/neondatabase/neon/pull/8050. We persist this incorrect image in the page cache
5. The basebackup requests come in. We map it to a `get_vectored` which doesn't use the materialized page cache.
This results in the correct page image which differs from the incorrect we cached at step (4). Hence, the assertion.

This sequence of events feels unlikely though. I'd say we shelve this for now and if it happens again we investigate immediately (or at least save the tenant, timeline and lsn).
